### PR TITLE
Dosdebug 34

### PR DESCRIPTION
**Pull request by andrewbird**

### CODE REVIEW

File: src/plugin/debugger/mhpdbgc.c [hunk 1]
```diff
@@ -817,31 +817,38 @@ static int is_valid_program_name(const char *s)
...
-static const char *get_mcb_name2(uint16_t seg, uint16_t off)
+static const char *get_name_from_mcb(struct MCB *mcb)
 {
-  char *target = MK_FP32(seg, off);
   const char *dos = "DOS", *fre = "FREE";
   static char name[9];
+
+  if (mcb->owner_psp == 0)
+    return fre;
+  if (mcb->owner_psp == 8)
```
Comment (stsp) [inline src/plugin/debugger/mhpdbgc.c]: Hmm, I've seen this 8 in freedos, but are
you sure its standard?

File: src/plugin/debugger/mhpdbgc.c [hunk 2]
```diff
@@ -817,31 +817,38 @@ static int is_valid_program_name(const char *s)
...
+
+static const char *get_mcb_name_walk_chain(uint16_t seg, uint16_t off)
+{
+  char *start, *end, *target = MK_FP32(seg, off);
   struct MCB *mcb;
 
   if (!lol)
     return NULL;
 
   for (mcb = MK_FP32(READ_WORD(lol - 2), 0); mcb->id == 'M'; /* */) {
```
Comment (stsp) [inline src/plugin/debugger/mhpdbgc.c]: Are you also detecting infinite loops here?

File: src/plugin/debugger/mhpdbgc.c [hunk 3]
```diff
@@ -954,6 +956,29 @@ static void mhp_ivec(int argc, char *argv[])
...
+
+  if (!lol) {
+    mhp_printf("DOS's LOL not set\n");
+    return;
+  }
+
+  mhp_printf("ADDR      PARAS  OWNER\n");
+  for (seg = READ_WORD(lol - 2), mcb = MK_FP32(seg, 0); mcb->id == 'M'; /* */) {
+    mhp_printf("%04x:0000 0x%04x [%s]\n", seg, mcb->size, get_name_from_mcb(mcb));
+    seg += (1 + mcb->size);
```
Comment (stsp) [inline src/plugin/debugger/mhpdbgc.c]: Yes, that's the right way of doing that. :)

File: src/plugin/debugger/mhpdbgc.c (hunk 1 above)
Comment (andrewbird) [reply src/plugin/debugger/mhpdbgc.c]: Yep, tested okay in PC-DOS 3.00, MS-DOS 6.22, DR-DOS 6.00, FreeDOS 1.20 and FDPP

File: src/plugin/debugger/mhpdbgc.c [hunk 4]
```diff
@@ -954,6 +956,29 @@ static void mhp_ivec(int argc, char *argv[])
...
+    mcb = MK_FP32(seg, 0);
+  }
+  if (mcb->id == 'Z') {
+    mhp_printf("%04x:0000 END\n", seg);
+  } else {
+    mhp_printf("MCB chain corrupt - missing final entry\n");
```
[the remark this answers is not in the document]
Comment (andrewbird) [reply src/plugin/debugger/mhpdbgc.c]: In can add a maximum loop counter here if you like, but what's the condition that could cause it, an MCB with negative size?

**Comment by stsp** [comment]
OK, thanks!
The size indeed seems to be unsigned.
It looks like dosdebug gets more attention
than fdpp these days, though. :)

**Comment by andrewbird** [comment]
Thanks.

**Comment by stsp** [comment]
> Thanks.

No, it was a hint. :)

**Comment by andrewbird** [comment]
Thanks for applying, I meant.

**Pull request merged**